source/opt/optimizer.cpp

```diff
@@ -201,12 +206,219 @@ Optimizer& Optimizer::RegisterSizePasses() {
       .RegisterPass(CreateAggressiveDCEPass());
 }
 
+bool Optimizer::RegisterPassesFromFlags(const std::vector<std::string>& flags) {
+  for (const auto& flag : flags) {
+    if (flag.size() < 2 || flag.substr(0, 2) != "--") {
+      Errorf(consumer(), nullptr, {},
+             "%s is not a valid flag.  Flag passes should have the form "
+             "'--pass_name[=pass_args]'",
+             flag.c_str());
+    }
+
+    std::unique_ptr<Optimizer::PassToken> p = GetPassFromFlag(flag);
+    if (p == nullptr) {
+      Errorf(consumer(), nullptr, {},
+             "flag %s does not specify a known SPIR-V optimizer flag",
+             flag.c_str());
+      return false;
+    }
+
+    RegisterPass(std::move(*p));
+  }
+
+  return true;
+}
+
+namespace {
+
+// Split the string |flag|, of the form '--pass_name[=pass_args]' into two
+// strings "pass_name" and "pass_args".  If |flag| has no arguments, the second
+// string will be empty.
+std::pair<std::string, std::string> SplitFlagArgs(const std::string& flag) {
+  assert(flag.find("--") == 0 && "ill-formed flag. It should start with '--'.");
+  size_t ix = flag.find('=');
+  return (ix != std::string::npos)
+             ? make_pair(flag.substr(2, ix - 1), flag.substr(ix + 1))
+             : make_pair(flag, std::string());
+}
+}  // namespace
+
+std::unique_ptr<Optimizer::PassToken> Optimizer::GetPassFromFlag(
```

Comment:
Could this be changed to register the pass? Then we could accept -O, -Os and --legalize-hlsl in the config file.